ddb-popper/ddb-popper.js

```diff
@@ -0,0 +1,200 @@
+        const windowCloseBtn = windowHeader.find(".close");
+
+        /**
+         * jquery reference to the D&D Beyond button to add to the sheet
+         */
+        const ddbButton = $(
+            `<a class="ddb-popup">
+                <img src=${DDBPopper.CONFIG.ddbLogo} style=${DDBPopper.CONFIG.imgStyle}> DDB
+            </a>`
+        );
+        
+        /**
+         * Remove any existing instances of the button on the sheet
+         */
+        windowHeader.find('.ddb-popup').remove();
+        windowCloseBtn.before(ddbButton);
+
+        /**
+         * When the DDB button is clicked, lookup the DDB URL, then look for an existing popup,
+         * and if none exists, and the DDB URL is null, open the form to set the URL, 
+         * otherwise if the popup does not exist but the DDB URL is not null, create the popup
+         * or else simply focus the existing popup
+         */
+        ddbButton.click(async ev => {
+            ev.preventDefault();
+            const actorDDBURL = await this._getActorDDBURL(data);
+
+            if (actorDDBURL == null || actorDDBURL.length == 0){
+                new DDBURLEntryForm(actorDDBURL, data, {closeOnSubmit: true}).render(true);
+            } else if ((this.existingPopup == null || this.existingPopup.closed) && actorDDBURL.length > 0) {
+                this.existingPopup = window.open(actorDDBURL, DDBPopper.CONFIG.moduleName, DDBPopper.CONFIG.windowFeatures);
+            } else if (this.existingPopup && actorDDBURL.length > 0) {
+                this.existingPopup.focus();
+            }
+        });
+
+        /**
+         * When the DDB button is right-clicked, lookup the DDB URL for the actor,
+         * then open the form to set the URL 
+         */
+        ddbButton.contextmenu(async ev => {
+            ev.preventDefault();
+            const actorDDBURL = await this._getActorDDBURL(data);
+
+            new DDBURLEntryForm(actorDDBURL, data, {closeOnSubmit: true}).render(true);
+        });
+
+        /**
+         * When the Close button is pressed close any open popups
+         */
+        windowCloseBtn.click(ev => {
+            ev.preventDefault();
+
+            if(this.existingPopup && !this.existingPopup.closed){
+                this.existingPopup.close();
+            }
+        });
+    }
+}
+
+/**
+ * A simple form for entering a URL.
+ * If the DDB URL already exists for the Actor referenced in the Actor sheet,
+ * display that in the form
+ * @param {String} actorDDBURL -- the DDB URL for the Actor that owns the Actor Sheet
+ * @param {Object} data -- the actor data from the sheet
+ * @param {Object} options -- any options to set (including those on the FormApplication/Application super classes)
+ */
+class DDBURLEntryForm extends FormApplication {
+    constructor(actorDDBURL, data, options){
+        super(data, options);
+        this.data = data;
+        this.actorId = this.data.actor._id;
```

Comment:
Are these variables defined in the extended class?

You already have access to this.data so I wouldn't create another variable for actorID and rather just access it from this.data. it leave less room for error as you currently are having to make sure both variables are in sync.

Reply:
This was something I wrestled with, but since I call actor._id a few times, I thought it might be better to declare it rather than rely on deep dot notation. What do you think?

Reply:
If create a helper function and call it. The function can do the deep lookup.